src/Shared/Targets/Wrappers/AspNetBufferingTargetWrapper.cs

```diff
@@ -0,0 +1,311 @@
+
+        /// <summary>
+        /// Provides access to the current request HttpContext.
+        /// </summary>
+        /// <returns>HttpContextAccessor or <c>null</c></returns>
+        internal IHttpContextAccessor HttpContextAccessor
+        {
+            get => _httpContextAccessor ?? (_httpContextAccessor = RetrieveHttpContextAccessor());
+            set => _httpContextAccessor = value;
+        }
+
+        private IHttpContextAccessor RetrieveHttpContextAccessor()
+        {
+#if ASP_NET_CORE
+            return ServiceLocator.ResolveService<IHttpContextAccessor>(ResolveService<IServiceProvider>(), LoggingConfiguration);
+#else
+            return new DefaultHttpContextAccessor();
+#endif
+        }
+
+        /// <summary>
+        /// The Key for the buffer dictionary in the HttpContext.Items collection
+        /// </summary>
+        private static readonly object HttpContextItemsKey = new object();
+
+        /// <summary>
+        /// Must be called by the HttpModule/Middleware upon starting.
+        /// This creates the dictionary in the HttpContext.Items with the proper key.
+        /// </summary>
+        /// <param name="context"></param>
+        internal static void Initialize(HttpContextBase context)
+        {
+            if (context == null)
+            {
+                return;
+            }
+
+            var bufferDictionary = GetBufferDictionary(context);
+
+            // If the dictionary is missing, create that first
+            if (bufferDictionary == null)
+            {
+                SetBufferDictionary(context);
+            }
+        }
+
+        /// <summary>
+        /// Obtains a slot in the buffer dictionary for 'this' class instance
+        /// If that does not exist, that is created first
+        /// if the dictionary does not exist, that is created first
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        private Internal.LogEventInfoBuffer GetOrCreateRequestBuffer(HttpContextBase context) 
+        {
+            // If the context is missing, stop
+            if (context == null)
+            {
+                return null;
+            }
+
+            var bufferDictionary = GetBufferDictionary(context);
+
+            // If the dictionary is missing, stop
+            if (bufferDictionary == null)
+            {
+                return null;
+            }
+
+            // if the slot for this class instance is missing, create that first
+            if (bufferDictionary.ContainsKey(this))
+            {
+                return bufferDictionary[this];
+            }
+
+            lock (bufferDictionary)
+            {
+                if (!bufferDictionary.ContainsKey(this))
+                {
+                    bufferDictionary.Add(this,
+                        new Internal.LogEventInfoBuffer(BufferSize, GrowBufferAsNeeded, BufferGrowLimit));
+                }
+            }
+
+            return bufferDictionary[this];
```

Comment:
Can this be moved inside the `lock`. Maybe use `bufferDictionary.TryGetValue` instead of `ContainsKey`

Reply:
Done as requested